feature(function): add custom functions documentation for AWS Amplify-backend (#8176)
aws-amplify/docs: src/pages/[platform]/build-a-backend/functions/custom-functions/index.mdx

import { getCustomStaticPath } from '@/utils/getCustomStaticPath';

export const meta = {
  title: 'Custom functions',
  description:
    'Use another AWS Lambda runtimes like python, golang to perform tasks and customize workflows.',
  platforms: [
    'android',
    'angular',
    'flutter',
    'javascript',
    'nextjs',
    'react',
    'react-native',
    'swift',
    'vue'
  ]
};

export function getStaticPaths() {
  return getCustomStaticPath(meta.platforms);
}

export function getStaticProps(context) {
  return {
    props: {
      platform: context.params.platform,
      meta
    }
  };
}

AWS Amplify Gen 2 Functions are AWS Lambda functions that can be used to perform tasks and customize workflows in your Amplify app.Functions can be written in Node.js, Python, Go, or any other language supported by AWS Lambda:

> <Callout warning>
> 
> **Note:** Custom runtimes are not supported in Amplify Functions directly. If you need docker support to use a custom runtime for example in Python, Docker is not supported in Amplify Hosting and Amplify backend auto build. You shouldn't use docker build in your function.There is an example of how to use python in a lambda function in the without Docker section below.
> 
> </Callout>

Reviewer: ```suggestion
<Callout warning>

**Note:** [Fullstack Git-based environments](https://docs.amplify.aws/react/how-amplify-works/concepts/#fullstack-git-based-environments) do not support Docker for functions bundling out of the box.

</Callout>
```

```ts title="amplify/functions/say-hello/resource.ts"
import { defineFunction } from "@aws-amplify/backend";
import { DockerImage, Duration } from "aws-cdk-lib";
import { Code, Function, Runtime } from "aws-cdk-lib/aws-lambda";
import { execSync } from "child_process";
import * as path from "path";

export const sayHello = defineFunction(
  (scope) =>
    new Function(scope, "say-hello", {
      handler: "index.handler",
      runtime: Runtime.PYTHON_3_9, // or any other python version
      timeout: Duration.seconds(20), //  default is 3 seconds
      code: Code.fromAsset(functionDir, {
        bundling: {
          image: DockerImage.fromRegistry("dummy"),
          local: {
            tryBundle(outputDir: string) {
              execSync(
                `python3 -m pip install -r ${path.join(functionDir, "requirements.txt")} -t ${path.join(outputDir)} --platform manylinux2014_x86_64 --only-binary=:all:`
              );
              execSync(`rsync -rLv ${functionDir}/* ${path.join(outputDir)}`);
              return true;
            },
          },
        },
      }),
    })
);
```

> Next, create the corresponding handler file at `amplify/functions/say-hello/handler.ts`. This is where your function code will go.

Reviewer 2: Shouldn't this be `amplify/functions/say-hello/index.py` instead of `amplify/functions/say-hello/handler.ts`?

Author: You're right, it's typo, I'm fixing it

          "message": "Hello World",
      }),
  }
```

The handler file _must_ export a function named "handler". This is the entry point to your function. For more information on writing functions, refer to the [AWS documentation for Lambda function handlers using python](https://docs.aws.amazon.com/lambda/latest/dg/python-handler.html).

If you need python packages, you can add them to a `requirements.txt` file in the same directory as your handler file. The `bundling` option in the `Code.fromAsset` method will install these packages for you.
Create a `requirements.txt` file in the same directory as your handler file. This file should contain the names of the packages you want to install. For example:

```txt title="amplify/functions/say-hello/requirements.txt"
request==2.25.1
some-other-package>=1.0.0
```

You're now ready to deploy your python function. Next is the same process as the Node.js/TypeScript function. Go to [Common steps for all languages](#common-steps-for-all-languages) to continue.

# Go functions
To get started, install go package from npm. go to your backend directory, you should see amplify folder inside it. and run the following command:

> ```npm i @aws-cdk/aws-lambda-go-alpha```

Reviewer: We should refrain from using alpha packages in docs.

We should:
- either convert this sample to use `Code.from..` and plain `Function` (like in case of python).
- or just remove Go sample.

Author: Here is an example of how this can be implemented using `Function`:

```
import { defineFunction } from "@aws-amplify/backend";
import { DockerImage, Duration } from "aws-cdk-lib";
import { Code, Function, Runtime } from "aws-cdk-lib/aws-lambda";
import { execSync } from "child_process";
import * as path from "path";
import { fileURLToPath } from "url";

const functionDir = path.dirname(fileURLToPath(import.meta.url));

export const sendSmsGoFunctionHandler = defineFunction(
  (scope) =>
    new Function(scope, "SendSmsGo", {
      handler: "bootstrap",
      runtime: Runtime.PROVIDED_AL2023, // or any other python version
      timeout: Duration.seconds(3), //  default is 3 seconds
      code: Code.fromAsset(functionDir, {
        bundling: {
          image: DockerImage.fromRegistry("dummy"),
          local: {
            tryBundle(outputDir: string) {
              execSync(`rsync -rLv ${functionDir}/* ${path.join(outputDir)}`);
              execSync(
                `cd ${path.join(outputDir)} && GOARCH=amd64 GOOS=linux go build -tags lambda.norpc -o ${path.join(outputDir)}/bootstrap ${functionDir}/main.go`
              );
              return true;
            },
          },
        },
      }),
    }),
);
```

```txt title="amplify/functions/say-hello/requirements.txt"
request==2.25.1
some-other-package>=1.0.0
```

You're now ready to deploy your python function. Next is the same process as the Node.js/TypeScript function. Go to [Common steps for all languages](#common-steps-for-all-languages) to continue.

## Go
To get started, Create a new directory and a resource file, `amplify/functions/say-hello/resource.ts`. Then, define the function with `defineFunction`:  

```ts title="amplify/functions/say-hello/resource.ts"
import { execSync } from "node:child_process";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import { defineFunction } from "@aws-amplify/backend";
import { DockerImage, Duration } from "aws-cdk-lib";
import { Code, Function, Runtime } from "aws-cdk-lib/aws-lambda";
import { defineFunction } from "@aws-amplify/backend";
import { DockerImage, Duration } from "aws-cdk-lib";
import { Code, Function, Runtime } from "aws-cdk-lib/aws-lambda";
export const sayHelloFunctionHandler = defineFunction(
  (scope) =>
    new Function(scope, "say-hello", {
      handler: "bootstrap",

>       runtime: Runtime.PROVIDED_AL2023, // or any other python version

Reviewer 3: ```suggestion
      runtime: Runtime.PROVIDED_AL2023,
```

>       timeout: Duration.seconds(3), //  default is 3 seconds

Reviewer 4: Are these comments still correct? They seem copied over from the previous Python example without change

Reviewer 3: ope, good call, added a suggestion to remove

Reviewer 4: Thanks. Also on the fence about mentioning default timeout value again, especially since in this example the timeout is already being set to the default value

Reviewer 3: I thought of removing it but left it for those that visit this page and jump straight to the Go example

Reviewer 4: I was thinking the same thing. Still potentially useful for that use case. Keeping it is fine with me

      }),
  }
```

The handler file _must_ export a function named "handler". This is the entry point to your function. For more information on writing functions, refer to the [AWS documentation for Lambda function handlers using Python](https://docs.aws.amazon.com/lambda/latest/dg/python-handler.html).

If you need Python packages, you can add them to a `requirements.txt` file in the same directory as your handler file. The `bundling` option in the `Code.fromAsset` method will install these packages for you.
Create a `requirements.txt` file in the same directory as your handler file. This file should contain the names of the packages you want to install. For example:

```txt title="amplify/functions/say-hello/requirements.txt"
request==2.25.1
some-other-package>=1.0.0
```

You're now ready to deploy your python function. Next is the same process as the Node.js/TypeScript function. Go to [Common steps for all languages](#common-steps-for-all-languages) to continue.

## Go
To get started, Create a new directory and a resource file, `amplify/functions/say-hello/resource.ts`. Then, define the function with `defineFunction`:  

```ts title="amplify/functions/say-hello/resource.ts"

> import { defineFunction } from "@aws-amplify/backend";
> import { DockerImage, Duration } from "aws-cdk-lib";
> import { Code, Function, Runtime } from "aws-cdk-lib/aws-lambda";
> import { execSync } from "node:child_process";
> import * as path from "node:path";
> import { fileURLToPath } from "node:url";

Reviewer 3: ```suggestion
import { execSync } from "node:child_process";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import { defineFunction } from "@aws-amplify/backend";
import { DockerImage, Duration } from "aws-cdk-lib";
import { Code, Function, Runtime } from "aws-cdk-lib/aws-lambda";
```

apologies, missed this one on reordering in my previous suggestions

    }),
);
```

Next, create the corresponding handler file at `amplify/functions/say-hello/main.go`. This is where your function code will go.

```go title="amplify/functions/say-hello/main.go"
package main

import (
	"context"
	"fmt"

	"github.com/aws/aws-lambda-go/lambda"
)

type Event struct {
	Arguments Arguments `json:"arguments"`
}

type Arguments struct {
	Title string `json:"phone"`
	Msg   string `json:"msg"`
}

func HandleRequest(ctx context.Context, event Event) (string, error) {
	fmt.Println("Received event: ", event)

	// fmt.Println("Message sent to: ", event.Arguments.Msg)
	// You can use lambda arguments in your code

	return "Hello World!", nil
}

func main() {
	lambda.Start(HandleRequest)
}
```

Then you should run the following command to build the go function:
```bash title="terminal" showLineNumbers={false}
go mod init lambda
```
then run to install the dependencies.

```bash title="terminal" showLineNumbers={false}
go mod tidy
```

You're now ready to deploy your golang function. Next is the same process as the Node.js/TypeScript function.

## Common steps for all languages  

Regardless of the language used, your function needs to be added to your backend.

> // highlight-next-line

Reviewer 3: ````suggestion
```ts title="amplify/backend.ts"
// highlight-next-line
````

import { sayHelloFunctionHandler } from './functions/say-hello/resource';

defineBackend({
  // highlight-next-line
  sayHelloFunctionHandler,
});
```

Now when you run `npx ampx sandbox` or deploy your app on Amplify, it will include your Function.

> To invoke your Function, we recommend adding your [Function as a handler for a custom query with your Amplify Data resource](/[platform]/build-a-backend/data/custom-business-logic/). This will enable you to strongly type Function arguments and the return statement, and use this to author your Function's business logic. To get started, open your `amplify/data/resource.ts` file and specify a new query in your schema:

Reviewer 3: ```suggestion
To invoke your function, we recommend adding your [function as a handler for a custom query with your Amplify Data resource](/[platform]/build-a-backend/data/custom-business-logic/). To get started, open your `amplify/data/resource.ts` file and specify a new query in your schema:
```

```ts title="amplify/data/resource.ts"

> Now when you run `npx ampx sandbox` or deploy your app on Amplify, it will include your function.

Reviewer 3: ```suggestion
```

import { getCustomStaticPath } from '@/utils/getCustomStaticPath';

export const meta = {
  title: 'Custom functions',
  description:

>     'Use another AWS Lambda runtimes like python, golang to perform tasks and customize workflows.',

Reviewer 3: ```suggestion
    'Use another AWS Lambda runtimes like Python, Golang to perform tasks and customize workflows.',
```

  platforms: [
    'android',
    'angular',
    'flutter',
    'javascript',
    'nextjs',
    'react',
    'react-native',
    'swift',
    'vue'
  ]
};

export function getStaticPaths() {
  return getCustomStaticPath(meta.platforms);
}

export function getStaticProps(context) {
  return {
    props: {
      platform: context.params.platform,
      meta
    }
  };
}

> AWS Amplify Gen 2 Functions are AWS Lambda functions that can be used to perform tasks and customize workflows in your Amplify app.Functions can be written in Node.js, Python, Go, or any other language supported by AWS Lambda:

Reviewer 3: ```suggestion
AWS Amplify Gen 2 functions are AWS Lambda functions that can be used to perform tasks and customize workflows in your Amplify app. Functions can be written in Node.js, Python, Go, or any [other language supported by AWS Lambda](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html).
```

<Callout warning>

**Note:** Custom runtimes are not supported in Amplify Functions directly. If you need docker support to use a custom runtime for example in Python, Docker is not supported in Amplify Hosting and Amplify backend auto build. You shouldn't use docker build in your function.There is an example of how to use python in a lambda function in the without Docker section below.

</Callout>

> Technically, you can use any language supported by AWS Lambda. These are NodeJS, Python, Java, .NET, Ruby.
> To use other languages in Lambda, such as Go or Rust, use an OS-only runtime.
> 

Reviewer 3: ```suggestion
```

Reviewer 4: Unresolving, as I think this still makes sense to remove. Is the mentioning of this relevant information on line 33 not sufficient?

> In this guide, you will learn how to create python and golang functions in Amplify Functions.

Reviewer 3: ```suggestion
In this guide, you will learn how to create Python and Go functions with Amplify functions.
```

> # Python functions
> To get started, create a new directory and a resource file, `amplify/functions/say-hello/resource.ts`. Then, define the Function with `defineFunction`:

Reviewer 3: ```suggestion
## Python

To get started, create a new directory and a resource file, `amplify/functions/say-hello/resource.ts`. Then, define the function with `defineFunction`:
```

```ts title="amplify/functions/say-hello/resource.ts"

> import { defineFunction } from "@aws-amplify/backend";
> import { DockerImage, Duration } from "aws-cdk-lib";
> import { Code, Function, Runtime } from "aws-cdk-lib/aws-lambda";
> import { execSync } from "child_process";
> import * as path from "path";
> import { fileURLToPath } from "url";

Reviewer 3: ```suggestion
import { execSync } from "node:child_process";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import { defineFunction } from "@aws-amplify/backend";
import { DockerImage, Duration } from "aws-cdk-lib";
import { Code, Function, Runtime } from "aws-cdk-lib/aws-lambda";
```

ordering nit for node imports

export const sayHelloFunctionHandler = defineFunction(
  (scope) =>
    new Function(scope, "say-hello", {
      handler: "index.handler",
      runtime: Runtime.PYTHON_3_9, // or any other python version
      timeout: Duration.seconds(20), //  default is 3 seconds
      code: Code.fromAsset(functionDir, {
        bundling: {
          image: DockerImage.fromRegistry("dummy"),
          local: {
            tryBundle(outputDir: string) {
              execSync(
                `python3 -m pip install -r ${path.join(functionDir, "requirements.txt")} -t ${path.join(outputDir)} --platform manylinux2014_x86_64 --only-binary=:all:`
              );
              execSync(`rsync -rLv ${functionDir}/* ${path.join(outputDir)}`);
              return true;
            },
          },
        },
      }),
    })
);
```

Next, create the corresponding handler file at `amplify/functions/say-hello/index.py`. This is where your function code will go.

```ts title="amplify/functions/say-hello/index.py"
import json
        
def handler(event, context):
  return {
      "statusCode": 200,
      "body": json.dumps({
          "message": "Hello World",
      }),
  }
```

> The handler file _must_ export a function named "handler". This is the entry point to your function. For more information on writing functions, refer to the [AWS documentation for Lambda function handlers using python](https://docs.aws.amazon.com/lambda/latest/dg/python-handler.html).

Reviewer 3: ```suggestion
The handler file _must_ export a function named "handler". This is the entry point to your function. For more information on writing functions, refer to the [AWS documentation for Lambda function handlers using Python](https://docs.aws.amazon.com/lambda/latest/dg/python-handler.html).
```

casing nit for proper noun

> If you need python packages, you can add them to a `requirements.txt` file in the same directory as your handler file. The `bundling` option in the `Code.fromAsset` method will install these packages for you.

Reviewer 3: ```suggestion
If you need Python packages, you can add them to a `requirements.txt` file in the same directory as your handler file. The `bundling` option in the `Code.fromAsset` method will install these packages for you.
```

Create a `requirements.txt` file in the same directory as your handler file. This file should contain the names of the packages you want to install. For example:

```txt title="amplify/functions/say-hello/requirements.txt"
request==2.25.1
some-other-package>=1.0.0
```

You're now ready to deploy your python function. Next is the same process as the Node.js/TypeScript function. Go to [Common steps for all languages](#common-steps-for-all-languages) to continue.

> # Go functions

Reviewer 3: ```suggestion
## Go
```

To get started, install go package from npm. go to your backend directory, you should see amplify folder inside it. and run the following command:

> ```npm i @aws-cdk/aws-lambda-go-alpha```

Reviewer 3: is this dependency still relevant?

>  create a new directory and a resource file, `amplify/functions/say-hello/resource.ts`. Then, define the Function with `defineFunction`:

Reviewer 3: ```suggestion
Create a new directory and a resource file, `amplify/functions/say-hello/resource.ts`. Then, define the function with `defineFunction`:
```

```ts title="amplify/functions/say-hello/resource.ts"

> import { defineFunction } from "@aws-amplify/backend";
> import { DockerImage, Duration } from "aws-cdk-lib";
> import { Code, Function, Runtime } from "aws-cdk-lib/aws-lambda";
> import { execSync } from "child_process";
> import * as path from "path";
> import { fileURLToPath } from "url";

Reviewer 3: ```suggestion
import { execSync } from "node:child_process";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import { defineFunction } from "@aws-amplify/backend";
import { DockerImage, Duration } from "aws-cdk-lib";
import { Code, Function, Runtime } from "aws-cdk-lib/aws-lambda";
```

            tryBundle(outputDir: string) {
              execSync(`rsync -rLv ${functionDir}/* ${path.join(outputDir)}`);
              execSync(
                `cd ${path.join(outputDir)} && GOARCH=amd64 GOOS=linux go build -tags lambda.norpc -o ${path.join(outputDir)}/bootstrap ${functionDir}/main.go`
              );
              return true;
            },
          },
        },
      }),
    }),
);
```

Next, create the corresponding handler file at `amplify/functions/say-hello/main.go`. This is where your function code will go.

```go title="amplify/functions/say-hello/main.go"
package main

import (
	"context"
	"fmt"

	"github.com/aws/aws-lambda-go/lambda"
)

type Event struct {
	Arguments Arguments `json:"arguments"`
}

type Arguments struct {
	Title string `json:"phone"`
	Msg   string `json:"msg"`
}

func HandleRequest(ctx context.Context, event Event) (string, error) {
	fmt.Println("Received event: ", event)

	// fmt.Println("Message sent to: ", event.Arguments.Msg)
	// You can use lambda arguments in your code

	return "Hello World!", nil
}

func main() {
	lambda.Start(HandleRequest)
}
```

Then you should run the following command to build the go function:

> ```bash title="terminal"

Reviewer 3: ````suggestion
```bash title="Terminal" showLineNumbers={false}
````

go mod init lambda
```
then run to install the dependencies.

> ```bash title="terminal"

Reviewer 3: ````suggestion
```bash title="Terminal" showLineNumbers={false}
````

go mod tidy
```

You're now ready to deploy your golang function. Next is the same process as the Node.js/TypeScript function.

> # Common steps for all languages

Reviewer 3: ```suggestion
## Common steps for all languages
```

> # Common steps for all languages
> Lastly, this function needs to be added to your backend.

Reviewer 3: ```suggestion
## Common steps for all languages

Regardless of the language used, your function needs to be added to your backend.
```

```ts title="amplify/backend.ts"
import { defineBackend } from '@aws-amplify/backend';
// highlight-next-line
import { sayHelloFunctionHandler } from './functions/say-hello/resource';

defineBackend({
  // highlight-next-line
  sayHelloFunctionHandler,
});
```

> Now when you run `npx ampx sandbox` or deploy your app on Amplify, it will include your Function.

Reviewer 4: Unresolving, as this seems like a change we'd want to make

> Now when you run `npx ampx sandbox` or deploy your app on Amplify, it will include your Function.

Author: Fixed

> To invoke your Function, we recommend adding your [Function as a handler for a custom query with your Amplify Data resource](/[platform]/build-a-backend/data/custom-business-logic/). This will enable you to strongly type Function arguments and the return statement, and use this to author your Function's business logic. To get started, open your `amplify/data/resource.ts` file and specify a new query in your schema:

Reviewer 3: can we reword this to remove the language around strongly typed handlers? that will be a typescript-specific feature

Author: Actually, it's just works as handler, it's important use case for Python or Go. It would be nice to have an example here.
